### PR TITLE
Wavelet denoising skeleton

### PR DESCRIPTION
**Pull request by Sand3r-**
The feature adds the skeleton of Wavelet denoising, containing just a 3 classes, adds 2 projects (libWavelet and libWavelet.Tests).

Beware,
Almost half of all the files (8) changed are just project files and related. 
I still fail to understand how much does the test/docstring boilerplate can increase the number of lines added. Over 900 lines of code that does nearly nothing :rofl:.

### CODE REVIEW

File: src/Spectre.libWavelet/Convolution.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+   limitations under the License.
+*/
+#pragma once
+#include <span.h>
+#include "DataTypes.h"
+
+namespace spectre::algorithm::wavelet
+{
+/// <summary>
+/// Filters the signal using rational transfer function.
```
Comment (gmrukwa) [inline src/Spectre.libWavelet/Convolution.h]: I am missing the *rational transfer function* part in the implementation. Where is it?

File: src/Spectre.libWavelet/PrecomputedDaubechiesCoefficients.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    -precomputedDaubechiesCoefficient[0],
+    precomputedDaubechiesCoefficient[1],
+    -precomputedDaubechiesCoefficient[2],
+    -precomputedDaubechiesCoefficient[3],
+    precomputedDaubechiesCoefficient[4],
+    precomputedDaubechiesCoefficient[5],
+    -precomputedDaubechiesCoefficient[6],
+    -precomputedDaubechiesCoefficient[7]
+};
+constexpr std::array<const DataType, 8> DecompositionYlNumerator{ // @gmrukwa: reverse of f0
```
Comment (gmrukwa) [inline src/Spectre.libWavelet/PrecomputedDaubechiesCoefficients.h]: I thinks that comments need to be updated.

File: src/Spectre.libWavelet/Spectre.libWavelet.vcxproj [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  </ImportGroup>
+  <ImportGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'" Label="PropertySheets">
+    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
+    <Import Project="..\PropertySheets\NativeProject.props" />
+  </ImportGroup>
+  <PropertyGroup Label="UserMacros" />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
+    <IntDir>$(Platform)$(Configuration)\</IntDir>
+  </PropertyGroup>
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
```
Comment (gmrukwa) [inline src/Spectre.libWavelet/Spectre.libWavelet.vcxproj]: Missing `TreatWarningAsError` setting in all configurations.

File: src/Spectre.libWavelet.Tests/Spectre.libWavelet.Tests.vcxproj [hunk 4]
```diff
@@ -0,0 +1,178 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\packages\opencv3.3.0.0.1\build\native\opencv3.props" Condition="Exists('..\packages\opencv3.3.0.0.1\build\native\opencv3.props')" />
```
Comment (gmrukwa) [inline src/Spectre.libWavelet.Tests/Spectre.libWavelet.Tests.vcxproj]: Why do we have OpenCV included in project file, but not in packages?

File: src/native-algorithms.sln [hunk 5]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
-VisualStudioVersion = 15.0.27130.2010
+VisualStudioVersion = 15.0.26730.12
```
Comment (gmrukwa) [inline src/native-algorithms.sln]: Unnecessary change.

File: src/Spectre.libWavelet.Tests/Spectre.libWavelet.Tests.vcxproj (hunk 4 above)
Comment (Sand3r-) [reply src/Spectre.libWavelet.Tests/Spectre.libWavelet.Tests.vcxproj]: Oh, it was just the leftover from the project stripping. Do you want me to remove it? It'll be added in the next PR.
And to answer the question precisely - I had to modify the project manually to match all the scripts used there, but forgot to add the opencv to the packages.

File: src/Spectre.libWavelet/Convolution.h (hunk 1 above)
Comment (Sand3r-) [reply src/Spectre.libWavelet/Convolution.h]: It's kind of, hard-coded. The denominator is set to one.

File: src/Spectre.libWavelet/Convolution.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Convolution"/> class.
+    /// </summary>
+    explicit Convolution();
+    /// <summary>
+    /// Performs filtering on the signal.
+    /// </summary>
+    /// <param name="signal">Signal to be filtered</param>
+    /// <returns>Filtered signal.</returns>
+    Signal Filter(const Signal& signal);
```
[the remark this answers is not in the document]
Comment (Sand3r-) [reply src/Spectre.libWavelet/Convolution.h]: To be honest I lost an idea myself on how shall I properly name this operation.

File: src/Spectre.libWavelet/PrecomputedDaubechiesCoefficients.h (hunk 2 above)
Comment (Sand3r-) [reply src/Spectre.libWavelet/PrecomputedDaubechiesCoefficients.h]: Does it? What changed since last time?

File: src/Spectre.libWavelet.Tests/Spectre.libWavelet.Tests.vcxproj (hunk 4 above)
Comment (gmrukwa) [reply src/Spectre.libWavelet.Tests/Spectre.libWavelet.Tests.vcxproj]: If it will be added soon, we may just add OpenCV to packages, and we're done ;) Just to keep these two in sync.

File: src/Spectre.libWavelet/Convolution.h (hunk 1 above)
Comment (gmrukwa) [reply src/Spectre.libWavelet/Convolution.h]: There was a convolution of the signal with itself somehow - fine. But no rational function parameters, even hard-coded are used there.

File: src/Spectre.libWavelet/PrecomputedDaubechiesCoefficients.h (hunk 2 above)
Comment (gmrukwa) [reply src/Spectre.libWavelet/PrecomputedDaubechiesCoefficients.h]: Naming of variables :P

**Comment by gmrukwa** [comment]
BTW. please update the version number before merging.

**Comment by gmrukwa** [comment]
[Implement wavelet denoising](https://trello.com/c/KzPr7RPc/205-implement-wavelet-denoising)

File: src/Spectre.libWavelet/PrecomputedDaubechiesCoefficients.h (hunk 2 above)
Comment (Sand3r-) [reply src/Spectre.libWavelet/PrecomputedDaubechiesCoefficients.h]: Oh, yeah... right

File: src/Spectre.libWavelet/Convolution.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (Sand3r-) [reply src/Spectre.libWavelet/Convolution.h]: Do you have any?

[the remark this answers is not in the document]
Comment (Sand3r-) [reply src/Spectre.libWavelet/Convolution.h]: Any ideas?

File: src/Spectre.libWavelet/Spectre.libWavelet.vcxproj (hunk 3 above)
Comment (Sand3r-) [reply src/Spectre.libWavelet/Spectre.libWavelet.vcxproj]: Kinda weird, the GUI says their on. Good catch

File: src/Spectre.libWavelet.Tests/ConvolutionTest.cpp [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        EXPECT_NO_THROW(Convolution());
+    }
+
+    class ConvolutionTest : public ::testing::Test
+    {
+    protected:
+        Convolution convolution;
+    };
+
+    TEST_F(ConvolutionTest, signal_filtering)
```
[the remark this answers is not in the document]
Comment (Sand3r-) [reply src/Spectre.libWavelet.Tests/ConvolutionTest.cpp]: No.

File: src/Spectre.libWavelet.Tests/MeanAbsoluteDeviationNoiseEstimatorTest.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    protected:
+        MeanAbsoluteDeviationNoiseEstimator estimator;
+    };
+
+    TEST_F(MeanAbsoluteDeviationNoiseEstimatorTest, estimate_noise)
+    {
+        Signal signal = { 1.0f, 2.0f, 3.0f };
+        constexpr DataType result = static_cast<DataType>(1.4650890114825907);
+        DataType estimate = estimator.Estimate(signal);
+        ASSERT_EQ(estimate, result);
```
[the remark this answers is not in the document]
Comment (Sand3r-) [reply src/Spectre.libWavelet.Tests/MeanAbsoluteDeviationNoiseEstimatorTest.cpp]: I have decided to use ```ASSERT_NEAR(3)```

**Comment by Sand3r-** [comment]
Hey, ```MeanAbsoluteDeviationNoiseEstimator``` was @gmrukwa 's idea, not mine :stuck_out_tongue_winking_eye: 

Thanks anyway!

File: src/Spectre.libWavelet.Tests/SoftThresholderTest.cpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    {
+    public:
+        SoftThresholderTest() : tresholder(1.0) {}
+    protected:
+        SoftThresholder tresholder;
+    };
+
+    TEST_F(SoftThresholderTest, properly_tresholds_signal)
+    {
+        Signal input = { 1.0f, 2.0f, 3.0f, 4.0f };
```
Comment (gmrukwa) [inline src/Spectre.libWavelet.Tests/SoftThresholderTest.cpp]: With such input we miss testing the property, that signal is bound to be nonnegative.

File: src/Spectre.libWavelet.Tests/packages.config [hunk 10]
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="gmock" version="1.7.0" targetFramework="native" />
+  <package id="Microsoft.Gsl" version="0.1.2.1" targetFramework="native" />
```
Comment (gmrukwa) [inline src/Spectre.libWavelet.Tests/packages.config]: We are still missing OpenCV in `packages.config`, to keep it in sync with `.vcxproj`.

File: src/Spectre.libWavelet.Tests/SoftThresholderTest.cpp (hunk 9 above)
Comment (Sand3r-) [reply src/Spectre.libWavelet.Tests/SoftThresholderTest.cpp]: Well, that's not true, nevertheless I'll modify the example so that it also contains the negative values.

File: src/Spectre.libWavelet.Tests/packages.config (hunk 10 above)
Comment (Sand3r-) [reply src/Spectre.libWavelet.Tests/packages.config]: The test project will not need it, hence I'll remove the OpenCV references from .```vcxproj```.
